core/controllers/base_test.py

```diff
@@ -1423,3 +1469,371 @@ def test_that_correct_derived_class_does_not_raise_exception(self):
         with self.swap(self, 'testapp', self.mock_testapp):
             self.post_json(
                 '/correctmock', payload, expected_status_int=200)
+
+
+class SchemaValidationIntegrationTests(test_utils.GenericTestBase):
+
+    non_schema_handlers = payload_validator.NON_SCHEMA_HANDLERS
+    wiki_page_link = (
+        'https://github.com/oppia/oppia/wiki/Validation-of-handler-args')
+
+    def test_every_handler_class_has_schema(self):
```

Comment:
In each of these tests there is a loop that iterates through all the routes in main.URLS and does various checks, instead place the checks in separate methods and add a single method which iterates through all the main.URLS and calls the other methods. This way we can reduce the repetition of code and reduce the time complexity also.

Reply:
Done, ptal.

Reply:
@rohitkatlaa 
Thank you for suggestion, I have a point to discuss,
Imagine a case when a developer missed to write a schema for any newly added handler, then one of our test functions will catch and raise errors somewhat like “Schema required in handlers [...]”. But the other test will raise errors like ‘'NoneType' object has no attribute 'keys', because BaseHandler is initialized with None. 
As the case when we divide the logic and check them differently, the user will get two error messages and he will know that the schema is required in any of the handlers.
But if we merge all the logic in a single test function, the user will only get a single error with 'NoneType' object that has no attribute 'keys', because all the assert statements are at the end of the test function and the **error msg is not very helpful for developers**.
Looking forward to your suggestions.

Reply:
Yes I agree, so I think we should be able to add the sufficient conditions before checking for the other tests, so that the key errors are not raised.

Reply:
Done, ptal